File: javascript/bob.js
/* eslint-disable no-unused-vars */
//
// This is only a SKELETON file for the 'Bob' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

function isQuestion(str) {
  return str[str.length - 1] === '?';
}

function filterOutNonAlpha(str) {
  return str.replace(/[^a-zA-Z]/g, '');
}

function isUpper(char) {
  return char.toUpperCase() === char;
}

function isTrue(c) {
  return c === true;
}

function isAllUpper(str) {
  str = filterOutNonAlpha(str);
  if (str === '') return false;
  return [...str].map(isUpper).every(isTrue);
}

export const hey = (message) => {
  message = message.trim();
  if (message === "") return "Fine. Be that way!";
  const isAQuestion = isQuestion(message);
  const isYelling = isAllUpper(message);
  if (isAQuestion && isYelling) return "Calm down, I know what I'm doing!";
  if (isAQuestion) return "Sure.";
  if (isYelling) return "Whoa, chill out!";
  return "Whatever.";
};

File: elixir/bob.ex
defmodule Bob do
  @spec hey(String.t()) :: String.t()
  def hey(input) do
    input = String.trim(input) 
    is_question = is_question?(input)
    is_empty = is_empty?(input)
    is_yelled = is_upper?(input)
    cond do 
      is_question and is_yelled -> "Calm down, I know what I'm doing!"
      is_question -> "Sure."
      not is_empty and is_yelled -> "Whoa, chill out!"
      is_empty?(input) -> "Fine. Be that way!"
      true -> "Whatever."
    end

  end


  def is_question?(str) do 
    String.last(str) == "?"
  end

  def is_upper?(str) do 
    filtered = remove_non_alpha(str)
    not is_empty?(filtered) and String.upcase(filtered) == filtered
  end 

  def is_empty?(str) do 
    str
    |> String.trim
    |> String.length 
    |> Kernel.==(0)
  end

  def remove_non_alpha(str) do 
    str 
    |> String.replace(~r/[^[:alpha:]]/, "")
  end

end
